File: ios/Classes/Layouts/FormExampleBuilder.swift
import UIKit
import GoogleMobileAds

/**
 * EXAMPLE CUSTOM BUILDER - Card Style với Gradient Background
 *
 * Đây là ví dụ hoàn chỉnh để bạn tham khảo và phát triển.
 *
 * Layout structure:
 * GADNativeAdView
 * └── UIView (Main Container với gradient background)
 *     ├── UIStackView (Content - Vertical)
 *     │   ├── UIView (Header)
 *     │   │   ├── GADMediaView (Full width với rounded corners)
 *     │   │   └── UIView (Icon overlay - bottom left của media)
 *     │   │       └── UIImageView (Icon - circular với border)
 *     │   ├── UIStackView (Info Section - với padding)
 *     │   │   ├── UILabel (Headline - large, bold)
 *     │   │   ├── UIStackView (Rating + Advertiser - horizontal)
 *     │   │   │   ├── StarRatingView (stars)
 *     │   │   │   └── UILabel (Advertiser)
 *     │   │   ├── UILabel (Body - description)
 *     │   │   └── UIStackView (Footer - Price + Store)
 *     │   │       ├── UILabel (Price - bold)
 *     │   │       └── UILabel (Store - subtle)
 *     │   └── UIButton (CTA - gradient button với shadow)
 *     └── UILabel (Ad Label - top right badge)
 *
 * Custom Features:
 * - Gradient background (blue to purple)
 * - Circular icon với white border overlay trên media
 * - Rounded corners cho tất cả elements
 * - CTA button với gradient và shadow effect
 * - Card elevation và shadow
 * - Ad badge ở góc trên phải
 */
enum FormExampleBuilder {

    // Custom dimensions
    private static let cardCornerRadius: CGFloat = 16
    private static let mediaHeight: CGFloat = 140  // Reduced from 200 to fit in 300px container
    private static let iconSize: CGFloat = 50      // Reduced from 60 to fit better
    private static let iconBorderWidth: CGFloat = 3
    private static let ctaCornerRadius: CGFloat = 24
    private static let cardPadding: CGFloat = 8
    private static let contentPadding: CGFloat = 12  // Reduced from 16 for tighter spacing

    static func build(styleManager: AdStyleManager) -> GADNativeAdView {
        let nativeAdView = GADNativeAdView()
        nativeAdView.translatesAutoresizingMaskIntoConstraints = false

        // Main container với gradient background
        let mainContainer = createMainContainer()

        // Content vertical stack
        let contentStack = createVerticalStack(spacing: 0, alignment: .fill)
        contentStack.isLayoutMarginsRelativeArrangement = true
        contentStack.layoutMargins = UIEdgeInsets(
            top: 0,
            left: 0,
            bottom: 0,
            right: 0
        )

        // === HEADER SECTION (Media + Icon Overlay) ===
        let (headerView, mediaView, iconView) = createHeaderSection()
        contentStack.addArrangedSubview(headerView)

        // === INFO SECTION (Headline, Rating, Body, Footer) ===
        let (infoStack, headlineLabel, ratingContainer, advertiserLabel, bodyLabel, priceLabel, storeLabel) = createInfoSection()
        contentStack.addArrangedSubview(infoStack)

        // === CTA BUTTON (wrapped in container for margins) ===
        let ctaButton = createCtaButton()
        let ctaContainer = UIView()
        ctaContainer.translatesAutoresizingMaskIntoConstraints = false
        ctaContainer.addSubview(ctaButton)

        NSLayoutConstraint.activate([
            ctaButton.topAnchor.constraint(equalTo: ctaContainer.topAnchor, constant: 0),
            ctaButton.leadingAnchor.constraint(equalTo: ctaContainer.leadingAnchor, constant: contentPadding),
            ctaButton.trailingAnchor.constraint(equalTo: ctaContainer.trailingAnchor, constant: -contentPadding),
            ctaButton.bottomAnchor.constraint(equalTo: ctaContainer.bottomAnchor, constant: -contentPadding),
            ctaButton.heightAnchor.constraint(equalToConstant: 48)
        ])

        contentStack.addArrangedSubview(ctaContainer)

        // Add content to main container
        mainContainer.addSubview(contentStack)

        // Add Ad Label badge (top right)
        let adBadge = createAdBadge()
        mainContainer.addSubview(adBadge)

        // Add to native ad view
        nativeAdView.addSubview(mainContainer)

        // Layout constraints
        setupConstraints(
            nativeAdView: nativeAdView,
            mainContainer: mainContainer,
            contentStack: contentStack,
            headerView: headerView,
            mediaView: mediaView,
            iconView: iconView,
            adBadge: adBadge,
            ctaButton: ctaButton
        )

        // Map views to GADNativeAdView
        mapViews(
            nativeAdView: nativeAdView,
            mediaView: mediaView,
            iconView: iconView,
            headlineLabel: headlineLabel,
            ratingContainer: ratingContainer,
            advertiserLabel: advertiserLabel,
            bodyLabel: bodyLabel,
            priceLabel: priceLabel,
            storeLabel: storeLabel,
            ctaButton: ctaButton
        )

        return nativeAdView
    }

    private static func createMainContainer() -> UIView {
        let container = UIView()
        container.translatesAutoresizingMaskIntoConstraints = false

        // Gradient background layer (blue to purple)
        let gradientLayer = CAGradientLayer()
        gradientLayer.colors = [
            UIColor(hex: "#667eea").cgColor, // Blue
            UIColor(hex: "#764ba2").cgColor  // Purple
        ]
        gradientLayer.startPoint = CGPoint(x: 0.5, y: 0)
        gradientLayer.endPoint = CGPoint(x: 0.5, y: 1)
        gradientLayer.cornerRadius = cardCornerRadius

        // Add gradient to container
        container.layer.insertSublayer(gradientLayer, at: 0)

        // Store gradient layer reference để update frame sau
        container.tag = 9999

        // Shadow effect
        container.layer.shadowColor = UIColor.black.cgColor
        container.layer.shadowOpacity = 0.2
        container.layer.shadowOffset = CGSize(width: 0, height: 4)
        container.layer.shadowRadius = 8

        return container
    }

    private static func createHeaderSection() -> (UIView, GADMediaView, UIImageView) {
        let headerView = UIView()
        headerView.translatesAutoresizingMaskIntoConstraints = false

        // Media View với rounded corners
        let mediaView = GADMediaView()
        mediaView.translatesAutoresizingMaskIntoConstraints = false
        mediaView.backgroundColor = UIColor(white: 0.94, alpha: 1)
        mediaView.layer.cornerRadius = cardCornerRadius
        mediaView.layer.maskedCorners = [.layerMinXMinYCorner, .layerMaxXMinYCorner]
        mediaView.clipsToBounds = true

        // Icon container (bottom left của media)
        let iconContainer = UIView()
        iconContainer.translatesAutoresizingMaskIntoConstraints = false

        // Icon - circular với white border
        let iconView = UIImageView()
        iconView.translatesAutoresizingMaskIntoConstraints = false
        iconView.contentMode = .scaleAspectFill
        iconView.backgroundColor = .white
        iconView.layer.cornerRadius = iconSize / 2
        iconView.layer.borderWidth = iconBorderWidth
        iconView.layer.borderColor = UIColor.white.cgColor
        iconView.clipsToBounds = true

        // Shadow cho icon
        iconView.layer.shadowColor = UIColor.black.cgColor
        iconView.layer.shadowOpacity = 0.3
        iconView.layer.shadowOffset = CGSize(width: 0, height: 2)
        iconView.layer.shadowRadius = 4
        iconView.layer.masksToBounds = false

        iconContainer.addSubview(iconView)
        headerView.addSubview(mediaView)
        headerView.addSubview(iconContainer)

        // Constraints cho header
        NSLayoutConstraint.activate([
            // Media fills header
            mediaView.topAnchor.constraint(equalTo: headerView.topAnchor),
            mediaView.leadingAnchor.constraint(equalTo: headerView.leadingAnchor),
            mediaView.trailingAnchor.constraint(equalTo: headerView.trailingAnchor),
            mediaView.heightAnchor.constraint(equalToConstant: mediaHeight),

            // Header height = media height
            headerView.heightAnchor.constraint(equalToConstant: mediaHeight),

            // Icon container bottom left
            iconContainer.leadingAnchor.constraint(equalTo: headerView.leadingAnchor, constant: contentPadding),
            iconContainer.bottomAnchor.constraint(equalTo: headerView.bottomAnchor, constant: 30), // Overlap
            iconContainer.widthAnchor.constraint(equalToConstant: iconSize),
            iconContainer.heightAnchor.constraint(equalToConstant: iconSize),

            // Icon fills container
            iconView.centerXAnchor.constraint(equalTo: iconContainer.centerXAnchor),
            iconView.centerYAnchor.constraint(equalTo: iconContainer.centerYAnchor),
            iconView.widthAnchor.constraint(equalToConstant: iconSize),
            iconView.heightAnchor.constraint(equalToConstant: iconSize)
        ])

        return (headerView, mediaView, iconView)
    }

    private static func createInfoSection() -> (UIStackView, UILabel, UIView, UILabel, UILabel, UILabel, UILabel) {
        let infoStack = createVerticalStack(spacing: 8, alignment: .fill)
        infoStack.isLayoutMarginsRelativeArrangement = true
        infoStack.layoutMargins = UIEdgeInsets(
            top: contentPadding + 30, // Extra space cho icon overlap
            left: contentPadding,
            bottom: contentPadding,
            right: contentPadding
        )

        // Headline - large & bold
        let headlineLabel = UILabel()
        headlineLabel.translatesAutoresizingMaskIntoConstraints = false
        headlineLabel.font = .systemFont(ofSize: 20, weight: .bold)
        headlineLabel.textColor = .white
        headlineLabel.numberOfLines = 2

        infoStack.addArrangedSubview(headlineLabel)

        // Rating + Advertiser row
        let ratingRow = createHorizontalStack(spacing: 8, alignment: .center)

        let ratingContainer = UIView()
        ratingContainer.translatesAutoresizingMaskIntoConstraints = false
        ratingContainer.isHidden = true
        ratingContainer.tag = 1001 // Tag để tìm sau

        let advertiserLabel = UILabel()
        advertiserLabel.translatesAutoresizingMaskIntoConstraints = false
        advertiserLabel.font = .systemFont(ofSize: 12)
        advertiserLabel.textColor = UIColor(white: 0.9, alpha: 1)
        advertiserLabel.isHidden = true

        ratingRow.addArrangedSubview(ratingContainer)
        ratingRow.addArrangedSubview(advertiserLabel)

        // Spacer để đẩy về trái
        let spacer1 = UIView()
        spacer1.translatesAutoresizingMaskIntoConstraints = false
        spacer1.setContentHuggingPriority(.defaultLow, for: .horizontal)
        ratingRow.addArrangedSubview(spacer1)

        infoStack.addArrangedSubview(ratingRow)

        // Body text
        let bodyLabel = UILabel()
        bodyLabel.translatesAutoresizingMaskIntoConstraints = false
        bodyLabel.font = .systemFont(ofSize: 14)
        bodyLabel.textColor = UIColor(white: 0.94, alpha: 1)
        bodyLabel.numberOfLines = 3
        bodyLabel.isHidden = true

        infoStack.addArrangedSubview(bodyLabel)

        // Footer - Price + Store
        let footerRow = createHorizontalStack(spacing: 12, alignment: .center)
        footerRow.isHidden = true

        let priceLabel = UILabel()
        priceLabel.translatesAutoresizingMaskIntoConstraints = false
        priceLabel.font = .systemFont(ofSize: 16, weight: .bold)
        priceLabel.textColor = .white
        priceLabel.isHidden = true

        let storeLabel = UILabel()
        storeLabel.translatesAutoresizingMaskIntoConstraints = false
        storeLabel.font = .systemFont(ofSize: 12)
        storeLabel.textColor = UIColor(white: 0.9, alpha: 1)
        storeLabel.isHidden = true

        footerRow.addArrangedSubview(priceLabel)
        footerRow.addArrangedSubview(storeLabel)

        // Spacer để đẩy về trái
        let spacer2 = UIView()
        spacer2.translatesAutoresizingMaskIntoConstraints = false
        spacer2.setContentHuggingPriority(.defaultLow, for: .horizontal)
        footerRow.addArrangedSubview(spacer2)

        infoStack.addArrangedSubview(footerRow)

        return (infoStack, headlineLabel, ratingContainer, advertiserLabel, bodyLabel, priceLabel, storeLabel)
    }

    private static func createCtaButton() -> UIButton {
        let ctaButton = UIButton(type: .system)
        ctaButton.translatesAutoresizingMaskIntoConstraints = false

        // Gradient background (pink to red)
        let gradientLayer = CAGradientLayer()
        gradientLayer.colors = [
            UIColor(hex: "#f093fb").cgColor, // Pink
            UIColor(hex: "#f5576c").cgColor  // Red
        ]
        gradientLayer.startPoint = CGPoint(x: 0, y: 0.5)
        gradientLayer.endPoint = CGPoint(x: 1, y: 0.5)
        gradientLayer.cornerRadius = ctaCornerRadius

        // Create container view cho gradient
        let gradientContainer = UIView()
        gradientContainer.layer.insertSublayer(gradientLayer, at: 0)
        gradientContainer.layer.cornerRadius = ctaCornerRadius
        gradientContainer.clipsToBounds = true

        // Store gradient layer để update frame sau
        ctaButton.tag = 8888

        // Text styling
        ctaButton.titleLabel?.font = .systemFont(ofSize: 16, weight: .bold)
        ctaButton.setTitleColor(.white, for: .normal)

        // Background color fallback
        ctaButton.backgroundColor = UIColor(hex: "#f5576c")
        ctaButton.layer.cornerRadius = ctaCornerRadius
        ctaButton.clipsToBounds = true

        // Shadow effect
        ctaButton.layer.shadowColor = UIColor.black.cgColor
        ctaButton.layer.shadowOpacity = 0.3
        ctaButton.layer.shadowOffset = CGSize(width: 0, height: 2)
        ctaButton.layer.shadowRadius = 4
        ctaButton.layer.masksToBounds = false

        return ctaButton
    }

    private static func createAdBadge() -> UILabel {
        let adBadge = UILabel()
        adBadge.translatesAutoresizingMaskIntoConstraints = false
        adBadge.text = "Ad"
        adBadge.font = .systemFont(ofSize: 10, weight: .bold)
        adBadge.textColor = UIColor(hex: "#333333")
        adBadge.backgroundColor = UIColor(hex: "#ffd700") // Gold
        adBadge.textAlignment = .center
        adBadge.layer.cornerRadius = 4
        adBadge.clipsToBounds = true

        // Padding
        adBadge.layoutMargins = UIEdgeInsets(top: 3, left: 6, bottom: 3, right: 6)

        // Shadow
        adBadge.layer.shadowColor = UIColor.black.cgColor
        adBadge.layer.shadowOpacity = 0.2
        adBadge.layer.shadowOffset = CGSize(width: 0, height: 1)
        adBadge.layer.shadowRadius = 2
        adBadge.layer.masksToBounds = false

        return adBadge
    }

    private static func setupConstraints(
        nativeAdView: GADNativeAdView,
        mainContainer: UIView,
        contentStack: UIStackView,
        headerView: UIView,
        mediaView: GADMediaView,
        iconView: UIImageView,
        adBadge: UILabel,
        ctaButton: UIButton
    ) {
        NSLayoutConstraint.activate([
            // Main container với margin
            mainContainer.topAnchor.constraint(equalTo: nativeAdView.topAnchor, constant: cardPadding),
            mainContainer.leadingAnchor.constraint(equalTo: nativeAdView.leadingAnchor, constant: cardPadding),
            mainContainer.trailingAnchor.constraint(equalTo: nativeAdView.trailingAnchor, constant: -cardPadding),
            mainContainer.bottomAnchor.constraint(equalTo: nativeAdView.bottomAnchor, constant: -cardPadding),

            // Content stack fills container
            contentStack.topAnchor.constraint(equalTo: mainContainer.topAnchor),
            contentStack.leadingAnchor.constraint(equalTo: mainContainer.leadingAnchor),
            contentStack.trailingAnchor.constraint(equalTo: mainContainer.trailingAnchor),
            contentStack.bottomAnchor.constraint(equalTo: mainContainer.bottomAnchor),

            // Ad badge (top right)
            adBadge.topAnchor.constraint(equalTo: mainContainer.topAnchor, constant: 12),
            adBadge.trailingAnchor.constraint(equalTo: mainContainer.trailingAnchor, constant: -12),
            adBadge.widthAnchor.constraint(greaterThanOrEqualToConstant: 30),
            adBadge.heightAnchor.constraint(equalToConstant: 20)
        ])

        // Update gradient layer frames khi layout thay đổi
        DispatchQueue.main.async {
            if let gradientLayer = mainContainer.layer.sublayers?.first as? CAGradientLayer {
                gradientLayer.frame = mainContainer.bounds
            }
        }
    }

    private static func mapViews(
        nativeAdView: GADNativeAdView,
        mediaView: GADMediaView,
        iconView: UIImageView,
        headlineLabel: UILabel,
        ratingContainer: UIView,
        advertiserLabel: UILabel,
        bodyLabel: UILabel,
        priceLabel: UILabel,
        storeLabel: UILabel,
        ctaButton: UIButton
    ) {
        nativeAdView.mediaView = mediaView
        nativeAdView.iconView = iconView
        nativeAdView.headlineView = headlineLabel
        nativeAdView.advertiserView = advertiserLabel
        nativeAdView.bodyView = bodyLabel
        nativeAdView.priceView = priceLabel
        nativeAdView.storeView = storeLabel
        nativeAdView.callToActionView = ctaButton

        // Rating sẽ được handle riêng khi ad load
    }
}

// MARK: - Helper Extensions

extension FormExampleBuilder {
    /// Create vertical stack view
    static func createVerticalStack(spacing: CGFloat, alignment: UIStackView.Alignment) -> UIStackView {
        let stack = UIStackView()
        stack.translatesAutoresizingMaskIntoConstraints = false
        stack.axis = .vertical
        stack.spacing = spacing
        stack.alignment = alignment
        stack.distribution = .fill
        return stack
    }

    /// Create horizontal stack view
    static func createHorizontalStack(spacing: CGFloat, alignment: UIStackView.Alignment) -> UIStackView {
        let stack = UIStackView()
        stack.translatesAutoresizingMaskIntoConstraints = false
        stack.axis = .horizontal
        stack.spacing = spacing
        stack.alignment = alignment
        stack.distribution = .fill
        return stack
    }
}

File: android/src/main/kotlin/com/tqc/ads/flutter_admob_native_ads/layouts/FormExampleBuilder.kt
package com.tqc.ads.flutter_admob_native_ads.layouts

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RatingBar
import android.widget.TextView
import com.google.android.gms.ads.nativead.MediaView
import com.google.android.gms.ads.nativead.NativeAdView
import com.tqc.ads.flutter_admob_native_ads.styling.AdStyleManager
import com.tqc.ads.flutter_admob_native_ads.utils.DimensionUtils

/**
 * EXAMPLE CUSTOM BUILDER - Card Style với Gradient Background
 *
 * Đây là ví dụ hoàn chỉnh để bạn tham khảo và phát triển.
 *
 * Layout structure:
 * NativeAdView
 * └── FrameLayout (Main Container với gradient background)
 *     ├── LinearLayout (Content - Vertical)
 *     │   ├── FrameLayout (Header)
 *     │   │   ├── MediaView (Full width với rounded corners)
 *     │   │   └── FrameLayout (Icon overlay - bottom left của media)
 *     │   │       └── ImageView (Icon - circular với border)
 *     │   ├── LinearLayout (Info Section - với padding)
 *     │   │   ├── TextView (Headline - large, bold)
 *     │   │   ├── LinearLayout (Rating + Advertiser - horizontal)
 *     │   │   │   ├── RatingBar (stars)
 *     │   │   │   └── TextView (Advertiser)
 *     │   │   ├── TextView (Body - description)
 *     │   │   └── LinearLayout (Footer - Price + Store)
 *     │   │       ├── TextView (Price - bold)
 *     │   │       └── TextView (Store - subtle)
 *     │   └── Button (CTA - gradient button với shadow)
 *     └── TextView (Ad Label - top right badge)
 *
 * Custom Features:
 * - Gradient background (blue to purple)
 * - Circular icon với white border overlay trên media
 * - Rounded corners cho tất cả elements
 * - CTA button với gradient và shadow effect
 * - Card elevation và shadow
 * - Ad badge ở góc trên phải
 */
object FormExampleBuilder {

    // Custom IDs
    private const val ID_MEDIA = 4001
    private const val ID_ICON = 4002
    private const val ID_HEADLINE = 4003
    private const val ID_RATING = 4004
    private const val ID_ADVERTISER = 4005
    private const val ID_BODY = 4006
    private const val ID_PRICE = 4007
    private const val ID_STORE = 4008
    private const val ID_CTA = 4009
    private const val ID_AD_LABEL = 4010

    // Custom dimensions
    private const val CARD_CORNER_RADIUS = 16f
    private const val MEDIA_HEIGHT_DP = 140f  // Reduced from 200 to fit in 300dp container
    private const val ICON_SIZE_DP = 50f      // Reduced from 60 to fit better
    private const val ICON_BORDER_WIDTH = 3f
    private const val CTA_CORNER_RADIUS = 24f
    private const val ELEVATION_DP = 8f

    fun build(context: Context, styleManager: AdStyleManager): NativeAdView {
        val nativeAdView = NativeAdView(context).apply {
            layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        }

        // Main container với gradient background
        val mainContainer = createMainContainer(context)

        // Content vertical layout
        val contentLayout = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        }

        // === HEADER SECTION (Media + Icon Overlay) ===
        val headerSection = createHeaderSection(context, styleManager)
        contentLayout.addView(headerSection)

        // === INFO SECTION (Headline, Rating, Body, Footer) ===
        val infoSection = createInfoSection(context, styleManager)
        contentLayout.addView(infoSection)

        // === CTA BUTTON ===
        val ctaButton = createCtaButton(context, styleManager)
        contentLayout.addView(ctaButton)

        // Add content to main container
        mainContainer.addView(contentLayout)

        // Add Ad Label badge (top right)
        val adLabel = createAdBadge(context)
        mainContainer.addView(adLabel)

        // Add to native ad view
        nativeAdView.addView(mainContainer)

        // Map views to NativeAdView
        mapViews(nativeAdView, headerSection, infoSection, ctaButton)

        return nativeAdView
    }

    private fun createMainContainer(context: Context): FrameLayout {
        return FrameLayout(context).apply {
            layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply {
                if (this is ViewGroup.MarginLayoutParams) {
                    setMargins(
                        DimensionUtils.dpToPx(context, 8f),
                        DimensionUtils.dpToPx(context, 8f),
                        DimensionUtils.dpToPx(context, 8f),
                        DimensionUtils.dpToPx(context, 8f)
                    )
                }
            }

            // Gradient background (blue to purple)
            background = GradientDrawable(
                GradientDrawable.Orientation.TOP_BOTTOM,
                intArrayOf(
                    Color.parseColor("#667eea"), // Blue
                    Color.parseColor("#764ba2")  // Purple
                )
            ).apply {
                cornerRadius = DimensionUtils.dpToPx(context, CARD_CORNER_RADIUS).toFloat()
            }

            // Card elevation & shadow
            elevation = DimensionUtils.dpToPx(context, ELEVATION_DP).toFloat()
        }
    }

    private fun createHeaderSection(context: Context, styleManager: AdStyleManager): FrameLayout {
        val headerContainer = FrameLayout(context).apply {
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        }

        // Media View với rounded corners
        val mediaView = MediaView(context).apply {
            id = ID_MEDIA
            layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                DimensionUtils.dpToPx(context, MEDIA_HEIGHT_DP)
            )
            // Rounded corners cho media
            clipToOutline = true
            outlineProvider = android.view.ViewOutlineProvider.BACKGROUND
            background = GradientDrawable().apply {
                cornerRadii = floatArrayOf(
                    DimensionUtils.dpToPx(context, CARD_CORNER_RADIUS).toFloat(),
                    DimensionUtils.dpToPx(context, CARD_CORNER_RADIUS).toFloat(),
                    DimensionUtils.dpToPx(context, CARD_CORNER_RADIUS).toFloat(),
                    DimensionUtils.dpToPx(context, CARD_CORNER_RADIUS).toFloat(),
                    0f, 0f, 0f, 0f
                )
                setColor(Color.parseColor("#f0f0f0"))
            }
        }

        // Icon overlay container (bottom left của media)
        val iconContainer = FrameLayout(context).apply {
            layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply {
                gravity = Gravity.BOTTOM or Gravity.START
                setMargins(
                    DimensionUtils.dpToPx(context, 16f),
                    0,
                    0,
                    DimensionUtils.dpToPx(context, -30f) // Overlap với section tiếp theo
                )
            }
        }

        // Icon - circular với white border
        val iconView = ImageView(context).apply {
            id = ID_ICON
            scaleType = ImageView.ScaleType.CENTER_CROP
            val size = DimensionUtils.dpToPx(context, ICON_SIZE_DP)
            layoutParams = FrameLayout.LayoutParams(size, size)

            // Circular mask với white border
            background = GradientDrawable().apply {
                shape = GradientDrawable.OVAL
                setColor(Color.WHITE)
                setStroke(
                    DimensionUtils.dpToPx(context, ICON_BORDER_WIDTH),
                    Color.WHITE
                )
            }
            clipToOutline = true
            outlineProvider = android.view.ViewOutlineProvider.BACKGROUND

            // Shadow
            elevation = DimensionUtils.dpToPx(context, 4f).toFloat()
        }

        iconContainer.addView(iconView)
        headerContainer.addView(mediaView)
        headerContainer.addView(iconContainer)

        return headerContainer
    }

    private fun createInfoSection(context: Context, styleManager: AdStyleManager): LinearLayout {
        val infoLayout = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            val padding = DimensionUtils.dpToPx(context, 12f)  // Reduced from 16 for tighter spacing
            setPadding(padding, padding, padding, padding)
        }

        // Headline - large & bold
        val headlineView = TextView(context).apply {
            id = ID_HEADLINE
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            textSize = 20f
            setTextColor(Color.WHITE)
            setTypeface(null, android.graphics.Typeface.BOLD)
            maxLines = 2
            ellipsize = android.text.TextUtils.TruncateAt.END
        }
        infoLayout.addView(headlineView)

        // Rating + Advertiser row
        val ratingRow = LinearLayout(context).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply {
                topMargin = DimensionUtils.dpToPx(context, 8f)
            }
        }

        val ratingBar = RatingBar(context, null, android.R.attr.ratingBarStyleSmall).apply {
            id = ID_RATING
            numStars = 5
            setIsIndicator(true)
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            visibility = View.GONE
        }

        val advertiserView = TextView(context).apply {
            id = ID_ADVERTISER
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply {
                marginStart = DimensionUtils.dpToPx(context, 8f)
            }
            textSize = 12f
            setTextColor(Color.parseColor("#e0e0e0"))
            visibility = View.GONE
        }

        ratingRow.addView(ratingBar)
        ratingRow.addView(advertiserView)
        infoLayout.addView(ratingRow)

        // Body text
        val bodyView = TextView(context).apply {
            id = ID_BODY
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply {
                topMargin = DimensionUtils.dpToPx(context, 12f)
            }
            textSize = 14f
            setTextColor(Color.parseColor("#f0f0f0"))
            maxLines = 3
            ellipsize = android.text.TextUtils.TruncateAt.END
            visibility = View.GONE
        }
        infoLayout.addView(bodyView)

        // Footer - Price + Store
        val footerRow = LinearLayout(context).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply {
                topMargin = DimensionUtils.dpToPx(context, 8f)
            }
            visibility = View.GONE
        }

        val priceView = TextView(context).apply {
            id = ID_PRICE
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            textSize = 16f
            setTextColor(Color.WHITE)
            setTypeface(null, android.graphics.Typeface.BOLD)
            visibility = View.GONE
        }

        val storeView = TextView(context).apply {
            id = ID_STORE
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply {
                marginStart = DimensionUtils.dpToPx(context, 12f)
            }
            textSize = 12f
            setTextColor(Color.parseColor("#e0e0e0"))
            visibility = View.GONE
        }

        footerRow.addView(priceView)
        footerRow.addView(storeView)
        infoLayout.addView(footerRow)

        return infoLayout
    }

    private fun createCtaButton(context: Context, styleManager: AdStyleManager): Button {
        return Button(context).apply {
            id = ID_CTA
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply {
                val margin = DimensionUtils.dpToPx(context, 16f)
                setMargins(margin, 0, margin, margin)
            }

            // Gradient background (yellow to orange)
            background = GradientDrawable(
                GradientDrawable.Orientation.LEFT_RIGHT,
                intArrayOf(
                    Color.parseColor("#f093fb"), // Pink
                    Color.parseColor("#f5576c")  // Red
                )
            ).apply {
                cornerRadius = DimensionUtils.dpToPx(context, CTA_CORNER_RADIUS).toFloat()
            }

            // Text styling
            textSize = 16f
            setTextColor(Color.WHITE)
            setTypeface(null, android.graphics.Typeface.BOLD)
            isAllCaps = false

            // Padding
            val paddingHorizontal = DimensionUtils.dpToPx(context, 24f)
            val paddingVertical = DimensionUtils.dpToPx(context, 12f)
            setPadding(paddingHorizontal, paddingVertical, paddingHorizontal, paddingVertical)

            // Shadow effect
            elevation = DimensionUtils.dpToPx(context, 4f).toFloat()
        }
    }

    private fun createAdBadge(context: Context): TextView {
        return TextView(context).apply {
            id = ID_AD_LABEL
            text = "Ad"
            layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply {
                gravity = Gravity.TOP or Gravity.END
                setMargins(
                    0,
                    DimensionUtils.dpToPx(context, 12f),
                    DimensionUtils.dpToPx(context, 12f),
                    0
                )
            }

            // Badge styling
            background = GradientDrawable().apply {
                setColor(Color.parseColor("#ffd700")) // Gold color
                cornerRadius = DimensionUtils.dpToPx(context, 4f).toFloat()
            }

            textSize = 10f
            setTextColor(Color.parseColor("#333333"))
            setTypeface(null, android.graphics.Typeface.BOLD)

            val padding = DimensionUtils.dpToPx(context, 6f)
            setPadding(padding, padding / 2, padding, padding / 2)

            elevation = DimensionUtils.dpToPx(context, 2f).toFloat()
        }
    }

    private fun mapViews(
        nativeAdView: NativeAdView,
        headerSection: FrameLayout,
        infoSection: LinearLayout,
        ctaButton: Button
    ) {
        // Map từ header
        nativeAdView.mediaView = headerSection.findViewById(ID_MEDIA)
        nativeAdView.iconView = headerSection.findViewById(ID_ICON)

        // Map từ info section
        nativeAdView.headlineView = infoSection.findViewById(ID_HEADLINE)
        nativeAdView.starRatingView = infoSection.findViewById(ID_RATING)
        nativeAdView.advertiserView = infoSection.findViewById(ID_ADVERTISER)
        nativeAdView.bodyView = infoSection.findViewById(ID_BODY)
        nativeAdView.priceView = infoSection.findViewById(ID_PRICE)
        nativeAdView.storeView = infoSection.findViewById(ID_STORE)

        // Map CTA
        nativeAdView.callToActionView = ctaButton
    }
}
